src/main/java/gregtech/common/metatileentities/electric/MetaTileEntityPump.java

```diff
@@ -212,10 +221,10 @@ private void checkFluidBlockAt(BlockPos pumpHeadPos, BlockPos checkPos) {
         boolean shouldCheckNeighbours = isStraightInPumpRange(checkPos);
 
         if (blockHere.getBlock() instanceof BlockLiquid ||
-            blockHere.getBlock() instanceof IFluidBlock) {
+                blockHere.getBlock() instanceof IFluidBlock) {
             IFluidHandler fluidHandler = FluidUtil.getFluidHandler(getWorld(), checkPos, null);
             FluidStack drainStack = fluidHandler.drain(Integer.MAX_VALUE, false);
-            if (drainStack != null && drainStack.amount > 0) {
+            if (drainStack != null && drainStack.amount > 0 && this.fluidFilter.testFluidStack(drainStack)) {
                 this.fluidSourceBlocks.add(checkPos);
             }
```

Comment:
To solve this the pump now has two dequeues : 
- `blockToCheck`  all neighbors that are further from the pump than the currently checked block are added to the dequeue
- `blockToReCheck` all neighbors closer to the pump than the checked block are added to this queue.
when checking block the first dequeue is emptied first using the older checking behavior, then the new dequeue get emptied but neighbors are added to the correct dequeue only if the current block is fully valid.

this way allow to infinitly pump fluids like water and ensure dequeues get empty over time on fluids like lava.
I didn't really find  a better way to achieve this but this way has the advantage of keeping older behavior as is in case some add-on makers rely on some of those functions.